system/jlib/jsocket.cpp

```diff
@@ -1098,12 +1173,63 @@ void CSocket::set_linger(int lingertime)
     }
 }
 
-void CSocket::set_keep_alive(bool set)
+void CSocket::setKeepAlive(bool set, int time, int intvl, int probes)
 {
     int on=set?1:0;
-    if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, (char*)&on, sizeof(on)) != 0) {
+    if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, (char*)&on, sizeof(on)) != 0)
+    {
         IWARNLOG("KeepAlive not set");
+        return;
+    }
+
+    if (!on)
+        return;
+
+    int srtn, optval;
+    socklen_t optlen = sizeof(optval);
+
+    optval = time;
+    srtn = setsockopt(sock, IPPROTO_TCP, TCP_KEEPIDLE, &optval, optlen);
+    if (srtn != 0)
+    {
+        IWARNLOG("KeepAlive time not set");
+        return;
+    }
+
+    optval = intvl;
+    srtn = setsockopt(sock, IPPROTO_TCP, TCP_KEEPINTVL, &optval, optlen);
+    if (srtn != 0)
+    {
+        IWARNLOG("KeepAlive probes not set");
+        return;
     }
+
+    optval = probes;
+    srtn = setsockopt(sock, IPPROTO_TCP, TCP_KEEPCNT, &optval, optlen);
+    if (srtn != 0)
+    {
+        IWARNLOG("KeepAlive probes not set");
+        return;
+    }
+}
+
+void CSocket::checkCfgKeepAlive()
+{
+    int time, intvl, probes;
+    if (queryKeepAlive(time, intvl, probes))
+        setKeepAlive(true, time, intvl, probes);
+}
+
+void CSocket::set_keep_alive(bool set)
+{
+    if (set)
+    {
+        int time, intvl, probes;
+        (void)queryKeepAlive(time, intvl, probes);
+        setKeepAlive(true, time, intvl, probes);
```

Comment:
Previously ISocket::set_keep_alive would only toggle the SO_KEEPALIVE option, now via setKeepAlive it will also make the system calls to set TCP_KEEPIDLE, TCP_KEEPINTVL and TCP_KEEPCNT , is that necessary/wanted?
If the global/export/keepalive settings are absent, it means it's going to force the keepalive idle/interval/probes to 7200,75,9- i.e. overridding the system defaults which may not be 7200,75.9

Why not keep as before and turn on keep alive only here (SO_KEEPALIVE)?
if global/expert/keepalive is configured, checkCfgKeepAlive() will set them before the socket is connect/pre_connect'd.

Reply:
Could acheive this (previous semantics of set_keep_alive) by coding it as:
```
void CSocket::set_keep_alive(bool set)
{
    setKeepAlive(set, 0, 0, 0);
}
```

and having setKeepAlive bail out before setting the idle/interval/probes if e.g. time==0

Reply:
I agree, and have changed the code so that you can set keepalive on and then it will not set time, intvl, probes unless you have also set those (non zero).
I assume zero is not a good value, we could make those -1 but I dont think its much of an issue.

Reply:
Right, probably -1 is better.

Reply:
changed